gen/internals.py

```diff
@@ -184,51 +221,203 @@ def add_entry(self, entry, replace_existing):
         # argument name.
         self._validate_by_arg = dict()
         self._multi_arg_validate = dict()
 
-        for fn in validate_fns:
-            parameters = get_function_parameters(fn)
+        for function in validate_functions:
+            parameters = get_function_parameters(function)
             # Could build up the single and multi parameter validation function maps in the same
             # thing but the timing / handling of when and how we run single vs. multi-parameter
             # validation functions is fairly different, the extra bit here simplifies the later code.
             if len(parameters) == 1:
-                self._validate_by_arg[parameters.pop()] = fn
-                assert not parameters
+                self._validate_by_arg.setdefault(parameters.pop(), list()).append(function)
             else:
-                self._multi_arg_validate[frozenset(parameters)] = fn
+                self._multi_arg_validate.setdefault(frozenset(parameters), list()).append(function)
+
+        for target in targets:
+            for parameter, function in target.yield_validates():
+                self._validate_by_arg.setdefault(parameter, list()).append(function)
+
+    def validate_single(self, name: str, value: str):
+        """Calls all validate functions which validate the given parameter name
+
+        The validate functions will raise an AssertionError which should be caught by the caller
+        when validation fails.
+        """
+        validate_fns = self._validate_by_arg.get(name)
+        if validate_fns is not None:
+            for validate_fn in validate_fns:
+                validate_fn(value)
+
+    # TODO(cmaloney): The distance between the validate_single and multi_arg_validate interface,
+    # while necessary for efficient functioning currently, is showing that there is tension between
+    # how / when each is run. Moving the multi-argument validation to as soon as possible after an
+    # argument set is finalized rather than one big pass at the end would likely make it much
+    # cleaner.
+    def yield_multi_argument_validate_errros(self, arguments: ArgumentDict):
+        for parameter_set, validate_fns in self._multi_arg_validate.items():
+            # Build up argument map for validate function. If any arguments are
+            # unset then skip this validate function.
+            kwargs = dict()
+            skip = False
+            for parameter in parameter_set:
+                # Exit early if the parameter was never calculated / asked for (We don't
+                # validate things that are never used or given)
+                if parameter not in arguments:
+                    skip = True
+                    break
+
+                # Exit if the parameter resolved to an error
+                resolvable = arguments[parameter]
+                if resolvable.is_error:
+                    skip = True
+                    break
```

Comment:
this is changing this logic, which is extending the scope of the code changed here, was trying to avoid changing this logic as much as possible because that means more things which could break inside this change / more diff to read / evaluate.